src/main/java/games/strategy/engine/data/RelationshipTracker.java

```diff
@@ -119,7 +120,7 @@ private RelationshipType getNullRelationshipType() {
      */
     @Override
     public int hashCode() {
-      return m_p1.hashCode() + m_p2.hashCode();
+      return Objects.hash(m_p1, m_p2);
```

Comment:
The Javadocs for this method state:

> override hashCode to make sure that each new instance of this class can be matched in the Hashtable even if it was put in as (p1,p2) and you want to get it out as (p2,p1).

The new implementation no longer has that property.  I'm assuming this is why you had to reverse the parameters in the call to `RelationshipTracker#isAllied()` in _BattleTracker.java_.  I would not suspect that the order of arguments to `isAllied()` would matter: `isAllied(p1, p2)` should yield the same result as `isAllied(p2, p1)`.  If the new `hashCode()` implementation precludes this, then it probably needs to be reverted.  Or, if you're still concerned about NPEs, then at least changed to

```java
Objects.hashCode(m_p1) + Objects.hashCode(m_p2)
```

Reply:
Will do...

Reply:
Thanks, @RoiEXLab.  Could you also revert the change to _BattleTracker.java_ so we can "prove" that the new implementation works? :smile: